src/ja/server/scheduler/algorithm.py

```diff
@@ -0,0 +1,115 @@
+from abc import ABC, abstractmethod
+from typing import List, Optional, Tuple
+from ja.common.job import Job
+from ja.server.database.database import ServerDatabase
+from ja.server.database.types.work_machine import WorkMachine
+
+
+class CostFunction(ABC):
+    """
+    A CostFunction is a collection of functions which determine the order in
+    which jobs are scheduled.
+    """
+
+    @abstractmethod
+    def calculate_cost(self, job: Job) -> int:
+        """!
+        Calculate the cost of a job.
+
+        The cost of a job is an integer. Higher values mean higher priority for
+        the job when considering which jobs to schedule.
+
+        @param job The job whose cost should be calculated.
+        @return The calculated cost of the job.
+        """
+
+    @abstractmethod
+    @property
+    def blocking_threshold(self) -> int:
```

Comment:
Why is blocking threshold a property of the cost function?
Shouldn't it be a property of the Scheduler?

Reply:
I wanted to make CostFunction control how jobs are scheduled (ordering and blocking/preempting policy). Maybe CostFunction is not a good name at all and that's why it is unclear.

Reply:
I think I understand now why you've made the thresholds part of the cost function.
The thresholds must be relative to the values produced by the cost function.
And because the specific values produced by a cost function depend on the implementation of the cost function, the thresholds should be bundled together with the cost function, correct?
I agree that the thresholds should be placed in the cost function.

Reply:
Yes, you're correct.